test_conformance/subgroups/test_workgroup_ballot.cpp

```diff
@@ -0,0 +1,973 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "procs.h"
+#include "subhelpers.h"
+#include "workgroup_common_templates.h"
+#include "harness/typeWrappers.h"
+#include <bitset>
+
+// Global/local work group sizes
+// Adjust these individually below if desired/needed
+#define GWS 2000
+#define LWS 200
+
+namespace {
+// Test for ballot functions
+template <typename Ty> struct BALLOT
+{
+    static void gen(Ty *x, Ty *t, cl_int *m, int ns, int nw, int ng)
+    {
+        int i, ii, j, k, n;
+        int nj = (nw + ns - 1) / ns;
+        int e;
+        ii = 0;
+        for (k = 0; k < ng; ++k)
+        { // for each work_group
+            for (j = 0; j < nj; ++j)
+            { // for each subgroup
+                ii = j * ns;
+                n = ii + ns > nw ? nw - ii : ns;
+                e = (int)(genrand_int32(gMTdata) % 3);
+
+                // Initialize data matrix indexed by local id and sub group id
+                switch (e)
+                {
+                    case 0: memset(&t[ii], 0, n * sizeof(Ty)); break;
+                    case 1:
+                        memset(&t[ii], 0, n * sizeof(Ty));
+                        i = (int)(genrand_int32(gMTdata) % (cl_uint)n);
+                        set_value(t[ii + i], 41);
+                        break;
+                    case 2: memset(&t[ii], 0xff, n * sizeof(Ty)); break;
+                }
+            }
+
+            // Now map into work group using map from device
+            for (j = 0; j < nw; ++j)
+            {
+                i = m[4 * j + 1] * ns + m[4 * j];
+                x[j] = t[i];
+            }
+
+            x += nw;
+            m += 4 * nw;
+        }
+    }
+
+    static int chk(Ty *x, Ty *y, Ty *mx, Ty *my, cl_int *m, int ns, int nw,
+                   int ng)
+    {
+        int ii, i, j, k, n;
+        int nj = (nw + ns - 1) / ns;
+        cl_uint4 tr, rr;
+
+        log_info("  sub_group_ballot...\n");
+
+        for (k = 0; k < ng; ++k)
+        { // for each work_group
+            // Map to array indexed to array indexed by local ID and sub group
+            for (j = 0; j < nw; ++j)
+            { // inside the work_group
+                i = m[4 * j + 1] * ns + m[4 * j];
+                mx[i] = x[j]; // read host inputs for work_group
+                my[i] = y[j]; // read device outputs for work_group
+            }
+
+            for (j = 0; j < nj; ++j)
+            { // for each subgroup
+                ii = j * ns;
+                n = ii + ns > nw ? nw - ii : ns;
+
+                // Check result
+                tr = { 0, 0, 0, 0 };
+                bs128 bs;
+                for (i = 0; i < n; ++i)
+                { // for each subgroup
+                    (mx[ii + i].s0 != 0) ? bs.set(i) : bs.reset(i);
+                }
+                // convert bs to uint4
+                auto const uint_mask = bs128{ static_cast<unsigned long>(-1) };
```

Comment:
If on a platform `unsigned long` is not 128 bits, this might not work. It would probably be safer to use `std::bitset::flip` or the `~` operator. This could be encapsulated in a free function if it's needed in a few places and you want to keep the `const`.

(untested)
```cpp
bs128 get_uint_mask() { bs128 mask; mask.flip(); return mask; }
```